### PR TITLE
Restrict replicating tables created on a cluster with higher version to a cluster with lower version

### PR DESCRIPTION
**Pull request by BaurzhanSakhariev**
Because older versions might not understand documents (5.4 won't able to understand OID-s) or metadata (copy_to -> sources changes as an example)

### CODE REVIEW

**Comment by BaurzhanSakhariev** [comment]
Replication from lower to higher is okay:
1. Prepare publisher with a lower version
```
cr8 run-crate 5.4.3
crash --host "localhost:4201"

select version['number'], port from sys.nodes;
+-------------------+-------------------------------------------------+
| version['number'] | port                                            |
+-------------------+-------------------------------------------------+
| 5.4.3             | {"http": 4201, "psql": 5435, "transport": 4301} |
+-------------------+-------------------------------------------------+
SELECT 1 row in set (0.011 sec)
```
```
cr> create table t (x int, o object(ignored) as (a int)) with (column_policy = 'dynamic');
CREATE OK, 1 row affected (1.611 sec)
cr> insert into t (x, o, dyn_empty_array) values (1, {"a" = 1, "2" = 2}, []);

cr> select _raw from t;
+------------------------------------------------+
| _raw                                           |
+------------------------------------------------+
| {"x":1,"o":{"a":1,"2":2},"dyn_empty_array":[]} |
+------------------------------------------------+

CREATE PUBLICATION pub1 FOR TABLE doc.t1;

create user subscriber with (password = 'secret');
GRANT DQL ON TABLE doc.t to subscriber;
```
2. replicate to a higher version:
```
cr8 run-crate branch:master 
select version['number'], port from sys.nodes;
+-------------------+-------------------------------------------------+
| version['number'] | port                                            |
+-------------------+-------------------------------------------------+
| 5.5.0             | {"http": 4200, "psql": 5434, "transport": 4300} |
+-------------------+-------------------------------------------------+
SELECT 1 row in set (0.008 sec)
cr> select * from t;
RelationUnknown[Relation 't' unknown] // replication not yet started

CREATE SUBSCRIPTION sub1 CONNECTION 'crate://localhost:4301?user=subscriber&password=secret&sslmode=disable' publication pub1;
CREATE OK, 1 row affected (0.197 sec)
cr> select * from t;
+---+------------------+
| x | o                |
+---+------------------+
| 1 | {"2": 2, "a": 1} |
+---+------------------+
SELECT 1 row in set (0.055 sec)
cr> select _raw from t;
+------------------------------------------------+
| _raw                                           |
+------------------------------------------------+
| {"x":1,"o":{"a":1,"2":2},"dyn_empty_array":[]} |
+------------------------------------------------+
SELECT 1 row in set (0.004 sec)
cr> insert into t (x) values (2);
OperationOnInaccessibleRelationException[The relation "doc.t" doesn't allow INSERT operations, because it is included in a logical replication subscription.]

```

**Comment by BaurzhanSakhariev** [comment]
Now replication from 5.5.0 to 5.4.3 which is not working
1. Prepare a publication on 5.5.0
```
create table t550 (x int, o object(ignored) as (a int)) with (column_policy = 'dynamic');
insert into t550(x, o, dyn_empty_array) values (1, {"a" = 1, "2" = 2}, []);
INSERT OK, 1 row affected (0.122 sec)
cr> select _raw from t550;
+------------------------------------------------------+
| _raw                                                 |
+------------------------------------------------------+
| {"1":1,"2":{"3":1,"_u_2":2},"_u_dyn_empty_array":[]} |
+------------------------------------------------------+
SELECT 1 row in set (0.850 sec)
cr> CREATE PUBLICATION pub1 FOR TABLE doc.t550;
CREATE OK, 1 row affected (0.085 sec)
cr> create user subscriber550 with (password = 'secret');
 
CREATE OK, 1 row affected (0.280 sec)

cr> GRANT DQL ON TABLE doc.t550 to subscriber550;
```

2. replicate to a lower version (**not working**)
```

select * from t550;
RelationUnknown[Relation 't550' unknown] // replication is not started yet

cr> CREATE SUBSCRIPTION sub1 CONNECTION 'crate://localhost:4300?user=subscriber550&password=secret&sslmode=disable' publication pub1;
CREATE OK, 1 row affected (0.121 sec)

cr> select * from  t550;
SQLParseException[Couldn't create execution plan from logical plan because of: the shard 2 of table [t550/b4QsxYvZQxerNEIZGpl7wg] is not available:
Collect[doc.t550 | [x, o] | true] (rows=unknown)]

cr> select _raw from t550;
SQLParseException[Couldn't create execution plan from logical plan because of: the shard 2 of table [t550/b4QsxYvZQxerNEIZGpl7wg] is not available:
Collect[doc.t550 | [_raw] | true] (rows=unknown)]
```

**Comment by BaurzhanSakhariev** [comment]
First commit failed with 
```
Warning, treated as error:
/var/lib/jenkins/workspace/CrateDB/crate_on_pr/docs/appendices/release-notes/5.5.0.rst:156:Failed to create a cross reference. A title or caption not found: memory.operation_limit
script returned exit code 2
```

I think it's because of underscore in the referenced name, faced it once in https://github.com/crate/crate/pull/13046#issuecomment-1262588202

File: docs/admin/logical-replication.rst [hunk 1]
```diff
@@ -30,6 +30,9 @@ Logical replication is useful for the following use cases:
 - Ensure high availability if one cluster becomes unavailable.
 
 - Replicating between different compatible versions of CrateDB.
+  Please note, that replicating from a higher version to a lower version
+  is not supported since old version might not be able to process data/metadata
+  of the newer version.
```
Comment (BaurzhanSakhariev) [inline docs/admin/logical-replication.rst:35]: Should we add some error on a subscribers side? 

Unfortunately, we cannot access Version of the publisher table without connecting to the remote cluster. It means we can end up with a not working subscription. 

If we decide to show an error, ideally we should drop subscription as well (and log a message explaining why).

Comment (seut) [reply docs/admin/logical-replication.rst:35]: Yes please, lets show an error and at best, throw the error while creating a subscription.

File: server/src/test/java/io/crate/integrationtests/LogicalReplicationITest.java [hunk 2]
```diff
@@ -257,6 +257,18 @@ public void test_subscribing_to_unknown_publication_raises_error() throws Except
...
+        // TODO: create a table with older version
+        executeOnPublisher("CREATE TABLE doc.t1 (id INT) WITH(" + defaultTableSettings() + ")");
+
+        createPublication("pub1", false, List.of("doc.t1"));
+        assertThatThrownBy(() -> executeOnSubscriber("CREATE SUBSCRIPTION sub1" +
+            " CONNECTION '" + publisherConnectionUrl() + "' publication pub1"))
+            .isExactlyInstanceOf(IllegalStateException.class)
+            .hasMessageContaining("One of the published tables has version higher than subscriber's minimal node version.");
+    }
+
```
Comment (BaurzhanSakhariev) [inline server/src/test/java/io/crate/integrationtests/LogicalReplicationITest.java:271]: We cannot do smth like CREATE TABLE  ... WITH(\"index.version.created\" = )", will create a unit tests with mocks instead and test this manually.

File: docs/appendices/release-notes/5.5.0.rst [hunk 3]
```diff
@@ -54,6 +54,9 @@ Breaking Changes
 - `UPDATE`` statements will continue on row failures instead of always showing 
   an error. The affected rows are displayed by the resulting row count.
 
+- Restricted :ref:`Logical Replication <administration-logical-replication>` of
+  tables created on a cluster with higher version to a cluster with lower version.
+
```
Comment (BaurzhanSakhariev) [inline docs/appendices/release-notes/5.5.0.rst:59]: There are some actually compatible versions where we didn't have indexing/metadata changes. If users used to do such replication, it's a breaking change for them.

File: docs/admin/logical-replication.rst (hunk 1 above)
Comment (BaurzhanSakhariev) [reply docs/admin/logical-replication.rst]: Added a check which throws an error on subscription creation.

File: server/src/test/java/io/crate/integrationtests/LogicalReplicationITest.java [hunk 4]
```diff
@@ -257,6 +260,35 @@ public void test_subscribing_to_unknown_publication_raises_error() throws Except
...
+        createPublication("pub1", false, List.of("doc.test"));
+        assertThatThrownBy(() -> executeOnSubscriber("CREATE SUBSCRIPTION sub1" +
+            " CONNECTION '" + publisherConnectionUrl() + "' publication pub1"))
+            .isExactlyInstanceOf(IllegalStateException.class)
+            .hasMessageContaining("One of the published tables has version higher than subscriber's minimal node version.");
+
+        var response = executeOnSubscriber("SELECT * FROM pg_subscription");
+        assertThat(response).hasRowCount(0);
+    }
+
```
Comment (BaurzhanSakhariev) [inline server/src/test/java/io/crate/integrationtests/LogicalReplicationITest.java:291]: One of 2 tests should be removed. I added an integration test as I also wanted to test that subscription is not created - or it's an obvious implication from unit test, showing that transport action throws an error?

**Comment by BaurzhanSakhariev** [comment]
unrelated bunch of io.crate.integrationtests.KillIntegrationTest failures

retest this please

File: docs/admin/logical-replication.rst [hunk 5]
```diff
@@ -30,6 +30,8 @@ Logical replication is useful for the following use cases:
 - Ensure high availability if one cluster becomes unavailable.
 
 - Replicating between different compatible versions of CrateDB.
+  Replicating tables created on a cluster with higher version to a cluster with
+  lower version is not supported.
```
Comment (seut) [inline docs/admin/logical-replication.rst:34]: higher hotfix versions should be supported
```suggestion
  Replicating tables created on a cluster with higher major/minor version to a cluster with
  lower major/minor version is not supported.
```

File: docs/appendices/release-notes/5.5.0.rst [hunk 6]
```diff
@@ -54,6 +54,9 @@ Breaking Changes
 - `UPDATE`` statements will continue on row failures instead of always showing 
   an error. The affected rows are displayed by the resulting row count.
 
+- Restricted :ref:`Logical Replication <administration-logical-replication>` of
+  tables created on a cluster with higher version to a cluster with lower version.
```
Comment (seut) [inline docs/appendices/release-notes/5.5.0.rst:58]: same here, clarify major/minor, hotfix versions should work
```suggestion
- Restricted :ref:`Logical Replication <administration-logical-replication>` of
  tables created on a cluster with higher version to a cluster with lower version.
```

File: server/src/main/java/io/crate/replication/logical/action/TransportCreateSubscriptionAction.java [hunk 7]
```diff
@@ -124,6 +144,14 @@ protected void masterOperation(CreateSubscriptionRequest request,
             );
     }
 
+    private static void checkVersionCompatibility(Version subscriberMinNodeVersion, Settings settings) {
+        Version publishedTableVersion = settings.getAsVersion(IndexMetadata.SETTING_VERSION_CREATED, null);
+        assert publishedTableVersion != null : "All published tables must have version created setting";
+        if (subscriberMinNodeVersion.before(publishedTableVersion)) {
```
Comment (seut) [inline server/src/main/java/io/crate/replication/logical/action/TransportCreateSubscriptionAction.java:150]: Please add a test verifying that subscribing to a higher hotfix version works.

```suggestion
        if (subscriberMinNodeVersion.beforeMajorMinor(publishedTableVersion)) {
```

File: server/src/test/java/io/crate/integrationtests/LogicalReplicationITest.java [hunk 8]
```diff
@@ -257,6 +260,35 @@ public void test_subscribing_to_unknown_publication_raises_error() throws Except
...
+        // Table to be replicated has a higher version subsriber's minimal node version.
+        publisherCluster.client()
+            .admin()
+            .indices()
+            .create(
+                new CreateIndexRequest("test")
+                .settings(
+                    Settings.builder()
+                        // Not necessary to set a valid version, only fact that version is higher than CURRENT is important for this test.
+                        .put("index.version.created", Version.CURRENT.internalId + 1)
```
Comment (seut) [inline server/src/test/java/io/crate/integrationtests/LogicalReplicationITest.java:275]:
```suggestion
                        .put("index.version.created", Version.CURRENT.internalId + 10000)
```

File: server/src/test/java/io/crate/replication/logical/action/TransportCreateSubscriptionActionTest.java [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+
+        RelationName relationName = RelationName.of(QualifiedName.of("t1"), Schemas.DOC_SCHEMA_NAME);
+        RelationMetadata relationMetadata = new RelationMetadata(
+            relationName,
+            List.of(
+                IndexMetadata.builder("t1")
+                .settings(
+                    // Not necessary to set a valid version, only fact that version is higher than CURRENT is important for this test.
+                    Settings.builder()
+                        .put(IndexMetadata.SETTING_VERSION_CREATED, Version.CURRENT.internalId + 1)
```
Comment (seut) [inline server/src/test/java/io/crate/replication/logical/action/TransportCreateSubscriptionActionTest.java:77]:
```suggestion
                        .put(IndexMetadata.SETTING_VERSION_CREATED, Version.CURRENT.internalId + 10000)
```

File: server/src/test/java/io/crate/integrationtests/LogicalReplicationITest.java (hunk 4 above)
Comment (seut) [reply server/src/test/java/io/crate/integrationtests/LogicalReplicationITest.java:291]: The transport action test should be enough. Testing that tables aren't created if the action throws any error is not related to this change.

File: server/src/test/java/io/crate/integrationtests/LogicalReplicationITest.java (hunk 8 above)
Comment (BaurzhanSakhariev) [reply server/src/test/java/io/crate/integrationtests/LogicalReplicationITest.java:275]: removed the itest

File: server/src/main/java/io/crate/replication/logical/action/TransportCreateSubscriptionAction.java (hunk 7 above)
Comment (BaurzhanSakhariev) [reply server/src/main/java/io/crate/replication/logical/action/TransportCreateSubscriptionAction.java]: Thanks added in https://github.com/crate/crate/pull/14869/commits/04832b31185d380b35cc40d37c460d9c32aa4242

File: server/src/test/java/io/crate/replication/logical/action/TransportCreateSubscriptionActionTest.java [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+        assertThatThrownBy(
+            () -> responseFuture.get()
+        )
+        .hasMessageContaining("One of the published tables has version higher than subscriber's minimal node version.");
+    }
+
+    @Test
+    public void test_subscribing_to_tables_with_higher_hotfix_works() throws Exception {
+        FutureActionListener<AcknowledgedResponse, AcknowledgedResponse> responseFuture = FutureActionListener.newInstance();
+        subscribeToTablesWithVersion(responseFuture, Version.CURRENT.internalId + 1);
```
Comment (seut) [inline server/src/test/java/io/crate/replication/logical/action/TransportCreateSubscriptionActionTest.java:74]: hotfix versions are increased by 100
```suggestion
        subscribeToTablesWithVersion(responseFuture, Version.CURRENT.internalId + 100);
```

Comment (BaurzhanSakhariev) [reply server/src/test/java/io/crate/replication/logical/action/TransportCreateSubscriptionActionTest.java]: thanks!

**Review by seut — APPROVED** [review]
👍

**Pull request merged**